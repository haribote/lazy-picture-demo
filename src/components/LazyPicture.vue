<template>
  <div class="picture" :class="classNames">
    <transition name="fade">
      <img v-if="!isLoaded" key="placeholder" :src="placeholderSrc" :alt="alt">
      <img v-else key="real" :src="alternativeSrc" :srcset="srcset" :sizes="sizes" :alt="alt">
    </transition>
  </div>
</template>

<script lang="ts">
require('intersection-observer')
import { Component, Prop, Vue } from 'vue-property-decorator'

@Component
export default class LazyPicture extends Vue {
  @Prop({ type: IntersectionObserver })
  observer = new IntersectionObserver(this.handleIntersection, {
    threshold: 0.5
  })

  @Prop({ required: true })
  placeholderSrc!: string

  @Prop({ required: true })
  srcset!: string

  @Prop({ required: true })
  sizes!: string

  @Prop({ default: '' })
  alt!: string

  @Prop({ default: false })
  isBlur!: boolean

  @Prop({ default: false })
  isLandscape!: boolean

  @Prop({ default: false })
  isPortrait!: boolean

  @Prop({ default: false })
  isSquare!: boolean

  isLoaded = false

  get alternativeSrc() {
    return this.srcset
      .split(',')[0]
      .replace(/^(.+\.(png|jpe?g|gif|webp)) +.+/, '$1')
  }

  get classNames() {
    const { isBlur, isLandscape, isPortrait, isSquare, isLoaded } = this

    return {
      'is-blur': isBlur && !isLoaded,
      'is-landscape': isLandscape,
      'is-portrait': isPortrait,
      'is-square': isSquare,
      'is-waiting': !isLoaded
    }
  }

  mounted() {
    this.observer.observe(this.$el)
  }

  preloadImage() {
    const { alternativeSrc, sizes, srcset } = this
    const imgEl = new Image()

    imgEl.onload = () => (this.isLoaded = true)

    imgEl.sizes = sizes
    imgEl.srcset = srcset
    imgEl.src = alternativeSrc
  }

  handleIntersection(entries: IntersectionObserverEntry[]) {
    entries.forEach(entry => {
      if (entry.isIntersecting) {
        this.observer.unobserve(this.$el)
        this.preloadImage()
      }
    })
  }
}
</script>

<style lang="scss" scoped>
.picture {
  position: relative;
  will-change: filter;

  img {
    will-change: opacity;
  }

  &.is-blur {
    transition: filter 300ms linear;
  }
}

.fade {
  @at-root {
    &-enter {
      opacity: 0;
    }
    &-enter-active {
      position: absolute;
      top: 0;
      left: 0;
      right: 0;
      margin: auto;
      transition: opacity 300ms linear;
    }
    &-enter-to {
      opacity: 1;
    }
    &-leave {
      opacity: 1;
    }
    &-leave-active {
      transition: opacity 300ms linear;
    }
    &-enter-to {
      opacity: 0.5;
    }
  }
}
</style>
