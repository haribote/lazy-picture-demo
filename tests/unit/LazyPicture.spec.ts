import 'intersection-observer'
import { shallowMount, Wrapper } from '@vue/test-utils'
import LazyPicture from '@/components/LazyPicture.vue'

describe('LazyPicture.vue', () => {
  let wrapper: Wrapper<LazyPicture>

  beforeEach(() => {
    wrapper = shallowMount(LazyPicture, {
      propsData: {
        observer: new IntersectionObserver(() => undefined),
        placeholderSrc:
          'https://c2.staticflickr.com/8/7019/6748869373_160471ded5_m.jpg',
        srcset:
          'https://c2.staticflickr.com/8/7019/6748869373_160471ded5_b.jpg 1x, https://c2.staticflickr.com/8/7019/6748869373_f7d856f99c_o.jpg 2x',
        sizes: '1024px',
        alt: 'image for testing'
      }
    })
  })

  afterEach(() => {
    wrapper.destroy()
  })

  describe('as default', () => {
    it('emits a ready event', () => {
      expect(wrapper.emitted('ready')[0]).toEqual([
        wrapper.vm.$el,
        (wrapper.vm as any)['preloadImage']
      ])
    })

    it('matches snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('as loaded', () => {
    it('matches snapshot', () => {
      wrapper.setData({ isLoaded: true })

      expect(wrapper.html()).toMatchSnapshot()
    })
  })
})
