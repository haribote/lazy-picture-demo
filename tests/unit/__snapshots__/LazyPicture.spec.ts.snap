// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LazyPicture.vue as default matches snapshot 1`] = `
<div class="picture is-waiting">
  <img src="https://c2.staticflickr.com/8/7019/6748869373_160471ded5_m.jpg" alt="image for testing" name="fade">
</div>
`;

exports[`LazyPicture.vue as loaded matches snapshot 1`] = `
<div class="picture">
  <img src="https://c2.staticflickr.com/8/7019/6748869373_160471ded5_b.jpg" srcset="https://c2.staticflickr.com/8/7019/6748869373_160471ded5_b.jpg 1x, https://c2.staticflickr.com/8/7019/6748869373_f7d856f99c_o.jpg 2x" sizes="1024px" alt="image for testing"
    name="fade">
</div>
`;
